wgengine/wgtun/tun.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package wgtun
+
+import (
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	"github.com/tailscale/wireguard-go/tun"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/filter"
+)
+
+const MaxPacketSize = device.MaxContentSize
+
+var (
+	ErrFiltered     = errors.New("packet dropped by filter")
+	ErrPacketTooBig = errors.New("packet too big")
+)
+
+// Device wraps a tun.Device from wireguard-go,
+// augmenting it with filtering and packet injection.
+// All the added work happens in Read and Write:
+// the other methods delegate to the underlying innerDevice.
+type Device struct {
+	innerDevice tun.Device
+
+	// Inbound packets can be written directly into the TUN device,
+	// but the other direction must wait on Wireguard to poll it.
+	// The packets are kept in the queue while waiting.
+	outbound chan []byte
+
+	// Filters are updated by ipn/local during the netmap refresh.
+	// They may be in use when this happens, hence the need for locking.
+	filterLock sync.Mutex
```

Comment:
`filterMu`, conventionally. Or really, just `mu`. Hopefully we won't need more than one mutex in this type.